gem/unconcatenate.py

```diff
@@ -0,0 +1,192 @@
+"""Utility functions for decomposing Concatenate nodes."""
+
+from __future__ import absolute_import, print_function, division
+from six.moves import range, zip
+
+from itertools import chain
+
+import numpy
+
+from gem.node import Memoizer, reuse_if_untouched
+from gem.gem import (ComponentTensor, Concatenate, FlexiblyIndexed,
+                     Index, Indexed, Node, reshape, view)
+from gem.optimise import remove_componenttensors
+
+
+__all__ = ['unconcatenate']
+
+
+def find_group(expressions):
+    """Finds a full set of indexed Concatenate nodes with the same
+    free index, if any such node exists.
+
+    Pre-condition: ComponentTensor nodes surrounding Concatenate nodes
+    must be removed.
+
+    :arg expressions: a multi-root GEM expression DAG
+    :returns: a list of GEM nodes, or None
+    """
+    free_indices = set().union(chain(*[e.free_indices for e in expressions]))
+
+    # Result variables
+    index = None
+    nodes = []
+
+    # Sui generis pre-order traversal so that we can avoid going
+    # unnecessarily deep in the DAG.
+    seen = set()
+    lifo = []
+    for root in expressions:
+        if root not in seen:
+            seen.add(root)
+            lifo.append(root)
+
+    while lifo:
+        node = lifo.pop()
+        if not free_indices.intersection(node.free_indices):
+            continue
+
+        if isinstance(node, Indexed):
+            child, = node.children
+            if isinstance(child, Concatenate):
+                i, = node.multiindex
+                assert i in free_indices
+                if (index or i) == i:
+                    index = i
+                    nodes.append(node)
+                    # Skip adding children
+                    continue
+
+        for child in reversed(node.children):
+            if child not in seen:
+                seen.add(child)
+                lifo.append(child)
+
+    return index and nodes
+
+
+def split_variable(variable_ref, index, multiindices):
+    """Splits a flexibly indexed variable along a concatenation index.
+
+    :param variable_ref: flexibly indexed variable to split
+    :param index: :py:class:`Concatenate` index to split along
+    :param multiindices: one multiindex for each split variable
+
+    :returns: generator of split indexed variables
+    """
```

Comment:
OK, so let's see if I understand what this is doing.  `variable_ref` is some kind of indexed object.  First I add shape to it in the axis order `index`, then all the other indices.

Now I want to pull apart the first dimension into sub-indexed bits, corresponding to the different parts of the Concatenate node I'm splitting?  So I index into a reshaped view of the component tensor.  But really I'm only slicing the first dimension, it's just you have to write this code because more general "numpy-like" slicing isn't implemented on component tensors?